File: android/app/src/main/java/com/blaineam/haven/core/PerceptualHash.kt
package com.blaineam.haven.core

import android.graphics.Bitmap

/**
 * A 64-bit fingerprint of what a picture LOOKS like, rather than of its bytes. Port of
 * `apple/HavenApp/PerceptualHash.swift` — the two must agree, since both platforms deduplicate the
 * same circle.
 *
 * Haven identifies media by sha-256 of the plaintext, which is right for storage and useless for
 * "is this the same photo?": the importer re-encodes everything it stages and re-encoding is not
 * reproducible, so importing the same archive twice gives every picture a different content hash.
 *
 * dHash: reduce to 9x8 grey, then one bit per adjacent horizontal pair — "is this pixel brighter
 * than the one to its right". What survives is coarse structure, which is what re-encoding preserves
 * and what a different photo does not share. Not a cryptographic hash; never use it as one.
 */
object PerceptualHash {

    /**
     * Bits that may differ before two pictures are considered different.
     *
     * 0 would demand bit-identical downsamples, which re-encoding does not guarantee; past ~10 it
     * starts matching photos that merely share a composition. A re-encode typically lands at 0-2 and
     * different pictures are almost always past 20, so 6 of 64 has wide margin on both sides.
     */
    const val DUPLICATE_THRESHOLD = 6

    const val WIDTH = 9
    const val HEIGHT = 8

    /** The pure half, so it can be tested on the JVM: `gray` is WIDTH*HEIGHT luminance samples. */
    fun dHashFromGray(gray: IntArray): ULong? {
        if (gray.size < WIDTH * HEIGHT) return null
        var hash = 0uL
        var bit = 0
        for (y in 0 until HEIGHT) {
            for (x in 0 until WIDTH - 1) {
                if (gray[y * WIDTH + x] > gray[y * WIDTH + x + 1]) hash = hash or (1uL shl bit)
                bit++
            }
        }
        return hash
    }

    /** Null when the bitmap cannot be reduced — callers must treat that as "unknown", never a match. */
    fun dHash(bitmap: Bitmap): ULong? {
        val small = runCatching { bitmap.scale(WIDTH, HEIGHT) }.getOrNull() ?: return null
        val px = IntArray(WIDTH * HEIGHT)
        small.getPixels(px, 0, WIDTH, 0, 0, WIDTH, HEIGHT)
        val gray = IntArray(px.size) { i ->
            val c = px[i]
            // Rec. 601 luma, integer — matches what a grayscale CGContext produces closely enough
            // for a 9x8 comparison, and the threshold absorbs the rest.
            ((c shr 16 and 0xFF) * 299 + (c shr 8 and 0xFF) * 587 + (c and 0xFF) * 114) / 1000
        }
        return dHashFromGray(gray)
    }

    private fun Bitmap.scale(w: Int, h: Int): Bitmap = Bitmap.createScaledBitmap(this, w, h, true)

    /** Hamming distance — how many of the 64 bits differ. */
    fun distance(a: ULong, b: ULong): Int = java.lang.Long.bitCount((a xor b).toLong())

    fun looksLikeTheSamePicture(a: ULong, b: ULong): Boolean = distance(a, b) <= DUPLICATE_THRESHOLD
}

File: apple/HavenApp/PerceptualHash.swift
import Foundation
import CoreGraphics

/// A 64-bit fingerprint of what a picture LOOKS like, rather than of its bytes.
///
/// Haven identifies media by sha-256 of the plaintext, which is exactly right for storage and
/// exactly useless for "is this the same photo?": the importer re-encodes everything it stages, and
/// re-encoding is not reproducible — a video is remuxed by AVAssetExportSession with its own
/// timestamps, posters are generated fresh, a JPEG round-trip need not be byte-identical. Import the
/// same archive twice and every content hash differs while every picture is the same. That is what
/// defeated the first duplicate sweep.
///
/// dHash (difference hash): reduce to 9×8 grey, then take one bit per adjacent horizontal pair —
/// "is this pixel brighter than the one to its right". What survives is the coarse structure of the
/// image, which is what re-encoding preserves and what a different photo does not share. It is
/// deliberately not a cryptographic hash and must never be used as one.
enum PerceptualHash {

    /// Bits that may differ before two pictures are considered different.
    ///
    /// 0 would demand bit-identical downsamples, which re-encoding does not guarantee. Anything past
    /// ~10 starts matching photos that merely share a composition — two frames of the same sunset,
    /// which are two posts. 6 of 64 is the usual working range for dHash and leaves a wide margin on
    /// both sides: a re-encode of the same JPEG typically lands at 0–2, and genuinely different
    /// pictures are almost always past 20.
    static let duplicateThreshold = 6

    private static let width = 9
    private static let height = 8

    /// Nil when the image cannot be rendered — callers must treat that as "unknown", never as a match.
    static func dHash(_ image: CGImage) -> UInt64? {
        var pixels = [UInt8](repeating: 0, count: width * height)
        guard let ctx = CGContext(data: &pixels, width: width, height: height,
                                  bitsPerComponent: 8, bytesPerRow: width,
                                  space: CGColorSpaceCreateDeviceGray(),
                                  bitmapInfo: CGImageAlphaInfo.none.rawValue) else { return nil }
        ctx.interpolationQuality = .medium
        ctx.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
        var hash: UInt64 = 0
        var bit = 0
        for y in 0..<height {
            for x in 0..<(width - 1) {
                if pixels[y * width + x] > pixels[y * width + x + 1] { hash |= (1 << UInt64(bit)) }
                bit += 1
            }
        }
        return hash
    }

    /// Hamming distance — how many of the 64 bits differ.
    static func distance(_ a: UInt64, _ b: UInt64) -> Int { (a ^ b).nonzeroBitCount }

    static func looksLikeTheSamePicture(_ a: UInt64, _ b: UInt64) -> Bool {
        distance(a, b) <= duplicateThreshold
    }
}
